fv/wireguard_test.go

```diff
@@ -0,0 +1,455 @@
+			cc.ExpectSome(w[1], w[0])
+			cc.CheckConnectivity()
+
+			for _, felix := range felixes {
+				Eventually(func() string {
+					// No tunnel implies un-encrypted communication.
+					out, _ := felix.ExecOutput("wg")
+					return out
+				}, "10s", "100ms").Should(BeEmpty())
+			}
+		})
+	})
+})
+
+var _ = infrastructure.DatastoreDescribe("WireGuard-Unsupported", []apiconfig.DatastoreType{apiconfig.EtcdV3, apiconfig.Kubernetes}, func(getInfra infrastructure.InfraFactory) {
+	var (
+		infra   infrastructure.DatastoreInfra
+		felixes []*infrastructure.Felix
+	)
+
+	BeforeEach(func() {
+		// Setup a single node cluster.
+		const nodeCount = 1
+
+		infra = getInfra()
+		felixes, _ = infrastructure.StartNNodeTopology(nodeCount, wireguardTopologyOptions(), infra)
+
+		// Install a default profile that allows all ingress and egress, in the absence of any Policy.
+		infra.AddDefaultAllow()
+
+		felixes[0].TriggerDelayedStart()
+	})
+
+	AfterEach(func() {
+		if CurrentGinkgoTestDescription().Failed {
+			felixes[0].Exec("ip", "link")
+			felixes[0].Exec("wg")
+		}
+
+		felixes[0].Stop()
+
+		if CurrentGinkgoTestDescription().Failed {
+			infra.DumpErrorData()
+		}
+		infra.Stop()
+	})
+
+	It("no Wireguard device exists", func() {
+		Eventually(func() string {
+			out, _ := felixes[0].ExecOutput("ip", "link", "show", wireguardInterfaceNameDefault)
+			return out
+		}, "10s", "100ms").Should(BeEmpty())
+	})
+
+	It("no wg tool info exists", func() {
+		Eventually(func() string {
+			out, _ := felixes[0].ExecOutput("wg")
+			return out
+		}, "10s", "100ms").Should(BeEmpty())
+	})
+})
+
+func wireguardTopologyOptions() infrastructure.TopologyOptions {
+	topologyOptions := infrastructure.DefaultTopologyOptions()
+
+	// For attaching tcpdump and waiting on calico-node to be ready.
+	topologyOptions.DelayFelixStart = true
+	// WireGuard doesn't support IPV6, disable it.
+	topologyOptions.EnableIPv6 = false
+	// Assigning workload IPs using IPAM api.
+	topologyOptions.IPIPRoutesEnabled = false
+	// Enable WireGuard.
+	felixConfig := api.NewFelixConfiguration()
+	felixConfig.SetName("default")
+	enabled := true
+	felixConfig.Spec.WireguardEnabled = &enabled
+	topologyOptions.InitialFelixConfiguration = felixConfig
+
+	// Debugging.
+	//topologyOptions.ExtraEnvVars["FELIX_DebugUseShortPollIntervals"] = "true"
+	//topologyOptions.FelixLogSeverity = "debug"
+
+	return topologyOptions
+}
+
+func enableDisableWireguard(client clientv3.Interface, enabled bool) {
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+	felixConfig, err := client.FelixConfigurations().Get(ctx, "default", options.GetOptions{})
+	Expect(err).NotTo(HaveOccurred())
+	felixConfig.Spec.WireguardEnabled = &enabled
+	felixConfig, err = client.FelixConfigurations().Update(ctx, felixConfig, options.SetOptions{})
+	Expect(err).NotTo(HaveOccurred())
+}
+
+func wireguardRuleRouteEntryExists(felix *infrastructure.Felix, exp *regexp.Regexp, entry string) bool {
```

Comment:
nit: RuleRouteEntry .... everywhere else you use RoutingRule - this makes it sound like something different.

Reply:
We are checking both RoutingRule and RouteEntry here, updated to `wireguardRoutingRuleRouteEntryExists`.

Reply:
This is a very odd interface.  You have a regex hardcoded into the method (which is fine for the test code), but then you pass in another regex to extract the priority.

What is exp?  Why does it need to be specified?

Also, there are now potentially multiple rules that will be installed.  This might be more useful if, given a "priority" and "table", it returns a list of from CIDRs.... or something like that.  Something that at least allows you to check multiple rules.  You can use it to check if rules have been deleted by checking for no CIDRs.  That kind of thing.